### PR TITLE
Feature/satellite graphs

### PR DESCRIPTION
**Pull request by maierlars**

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -329,16 +359,98 @@ exports.test = function (global) {
...
+
+    function createSmartGraph(name) {
+      var graph_module = require("@arangodb/smart-graph");
+      if (graph_module._exists(name)) {
+        return ;
+      }
+      let vertexCollectionName = name + "_vertex";
+      let edgesCollectionName = name + "_edge";
+
+      let opts = {smartGraphAttribute: "value2", numberOfShards: 3 };
```
Comment (goedderz) [inline test.js]:
```suggestion
      let opts = {smartGraphAttribute: "value2", numberOfShards };
```
There is a global `numberOfShards`.

File: test.js [hunk 2]
```diff
@@ -329,16 +359,98 @@ exports.test = function (global) {
...
+    let sg = createSmartGraph("smart");
+
+    let satg;
+    if (supportsSatelliteGraphs) {
+      print("Creating satellite graph");
+      satg = createSatelliteGraph("sat");
+    }
+
+    if (!gc || !sg || (supportsSatelliteGraphs && !satg)) {
+      return ;
```
Comment (goedderz) [inline test.js]: Don't you want to throw here?

File: test.js [hunk 3]
```diff
@@ -350,19 +462,54 @@ exports.test = function (global) {
...
+    function createEdges(n) {
+      fillGraphEdges(gc.edges, n, gc.vertex);
+      fillGraphEdges(sg.edges, n, sg.vertex);
+      if (supportsSatelliteGraphs) {
+        fillGraphEdges(satg.edges, n, satg.vertex);
+      }
+    }
+
     if (global.small) {
+      createVertexes(10000);
```
Comment (goedderz) [inline test.js]:
```suggestion
      createVertices(10000);
```
😱

File: test.js [hunk 4]
```diff
@@ -2200,15 +2370,67 @@ exports.test = function (global) {
...
+           params: {  func: genericSatelliteGraph,
+                      queryString: `
+                        for doc in @@c
+                          filter doc.value1 >= 0 and doc.value1 <= 10
+                          let vkey = CONCAT(@v,"/smart", doc.value3, ":test", doc.value3)
+                          for v, e, p in 1..4 outbound vkey graph @g
+                            return {doc, p}
+                        `,
+                    }
+        },
```
Comment (goedderz) [inline test.js]: One additional test, where a `DistributeNode` would distribute the lookup-values to the DBServers, would be nice, so we can later see the improvement when e.g. the `remove-satellite-joins` rule is extended to `DistributeNode`s.

File: test.js [hunk 5]
```diff
@@ -2557,6 +2779,65 @@ exports.test = function (global) {
...
+
+      if (global.medium) {
+        options.collections.push({ name: "values100000", label: "100k", size: 100000 });
+      }
+
+      if (global.big) {
+        options.collections.push({ name: "values1000000", label: "1000k", size: 1000000 });
+      }
+
+      /* We run each test case with splicing enabled and with splicing disabled */
```
Comment (goedderz) [inline test.js]: Do we? ;)

File: test.js (hunk 3 above)
Comment (maierlars) [reply test.js]: ... I know, but `Vertexes` is AE and arangodb uses AE. :(

**Comment by markuspf** [comment]
I know this is not entirely your fault, but I just spent 5 minutes even trying to run the performance test locally; 

I had to add `--javascript.module-directory $someprefix/enterprise/js` to the command that is given in the `README` so that I could run the tests.

It might be quite nice to have a simple bash script to which one can give the base path of the ArangoDB repository to test?

It might be of fairly low up-front (added)value but we might want to think about adding the other traversal variants as well (`SHORTEST_PATH`, `K_SHORTEST_PATHS`). I would hope the performance improvements for them to be in exactly the same ballpark, but maybe its better to check.

Generally this looks  good to me though.

**Pull request merged**